clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -2857,32 +2857,64 @@ static void handleWeakImportAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
 }
 
 // Checks correctness of mutual usage of different work_group_size attributes:
-// reqd_work_group_size, max_work_group_size. Values of reqd_work_group_size
-// arguments shall be equal or less than values coming from max_work_group_size.
+// reqd_work_group_size, max_work_group_size and max_global_work_dim.
+// Values of reqd_work_group_size arguments shall be equal or less than values
+// coming from max_work_group_size.
+// In case of 'max_global_work_dim' attribute equals to 0 we shall ensure
+// that if max_work_group_size and reqd_work_group_size attributes exist,
+// then they are equal (1, 1, 1).
 static bool checkWorkGroupSizeValues(Sema &S, Decl *D, const ParsedAttr &Attr,
                                      uint32_t WGSize[3]) {
+  bool Result = true;
+  if (Attr.getKind() == ParsedAttr::AT_SYCLIntelMaxGlobalWorkDim) {
+    if (const SYCLIntelMaxWorkGroupSizeAttr *A =
+            D->getAttr<SYCLIntelMaxWorkGroupSizeAttr>()) {
+      if (A->getXDim() != 1 || A->getYDim() != 1 || A->getZDim() != 1) {
+        S.Diag(A->getLocation(), diag::err_sycl_x_y_z_arguments_must_be_one)
+            << A << Attr;
+        Result = false;
+      }
+    }
+    if (const ReqdWorkGroupSizeAttr *A =
+            D->getAttr<ReqdWorkGroupSizeAttr>()) {
+      if (A->getXDim() != 1 || A->getYDim() != 1 || A->getZDim() != 1) {
+        S.Diag(A->getLocation(), diag::err_sycl_x_y_z_arguments_must_be_one)
+            << A << Attr;
+        Result = false;
+      }
+    }
+    return Result;
+  }
+  if (const SYCLIntelMaxGlobalWorkDimAttr *A =
+          D->getAttr<SYCLIntelMaxGlobalWorkDimAttr>()) {
+    if (A->getNumber() == 0) {
+      if (!(WGSize[0] == 1 && WGSize[1] == 1 && WGSize[2] == 1)) {
+        S.Diag(Attr.getLoc(), diag::err_sycl_x_y_z_arguments_must_be_one)
+            << Attr << A;
+        Result = false;
+      }
+    }
+  }
   if (const SYCLIntelMaxWorkGroupSizeAttr *A =
-      D->getAttr<SYCLIntelMaxWorkGroupSizeAttr>()) {
+          D->getAttr<SYCLIntelMaxWorkGroupSizeAttr>()) {
     if (!(WGSize[0] <= A->getXDim() && WGSize[1] <= A->getYDim() &&
           WGSize[2] <= A->getZDim())) {
       S.Diag(Attr.getLoc(), diag::err_conflicting_sycl_function_attributes)
           << Attr << A->getSpelling();
-      D->setInvalidDecl();
-      return false;
+      Result = false;
     }
   }
-
-  if (const ReqdWorkGroupSizeAttr *A = D->getAttr<ReqdWorkGroupSizeAttr>()) {
+  else if (const ReqdWorkGroupSizeAttr *A =
```

Comment:
Please add a Sema test for attributes' conflict (unless I've overlooked it). Also, why don't we want to diagnose two potential conflicts at once, as would've been the case with:
```suggestion
  if (const ReqdWorkGroupSizeAttr *A =
```
?

Reply:
Test was already done in a scope of a work for `max_work_group_size` attr.

Reply:
Ah, of course.

Reply:
Suggestion applied